[WIP] Refactor tasks to use hooks instead of inject

srallen opened the pull request:
_Please request review from `@zooniverse/frontend` team. If PR is related to design, please request review from `@beckyrother` in addition._ 

Package: lib-classifier

Describe your changes:
I wanted to get a start on writing a story for #1316 but got a bit sidetracked with some aspects of the task stories being broken. I was encountering this error when trying to use the knobs: https://github.com/mobxjs/mobx-react#the-set-of-provided-stores-has-changed-error

I read this to mean that the Tasks component should be updated to use context and hooks, which I did, but that didn't fix the error. The refactor is still worth keeping because we'll need to do this anyway for our connect components.

I also fixed the background color of the task area in the stories, fixed a key error on the task help, and added the ellipse to the drawing task story.

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?

## Components
- [ ] Has a storybook story been created or updated?
- [ ] Is the component accessible? 
  - [ ] Can it be used with a screen reader? [BBC guide to testing with VoiceOver](https://bbc.github.io/accessibility-news-and-you/accessibility-and-testing-with-voiceover-os.html)
  - [ ] Can it be used from the keyboard? [WebAIM guide to keyboard testing](https://webaim.org/techniques/keyboard/#testing)
  - [ ] Is it passing accessibility checks in the storybook?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `yarn panic && yarn bootstrap` or `docker-compose up --build` and app works as expected?

## Publishing

- [ ] Is the changelog updated?
- [ ] Are the dependencies updated for apps and libraries that are using the newly published library?

## Post-merging

- [ ] Did the app deploy to https://frontend.preview.zooniverse.org/projects/:project-name/:owner or https://frontend.preview.zooniverse.org/about?
- [ ] Is the new feature working or bug now fixed?
  - [ ] Is there a Talk or blog post written to announce the new feature(s)?
- [ ] Is the design working across browsers (Firefox, Chrome, Edge, Safari) and mobile?
  - [ ] Is this approved by our designer?
- [ ] Is this ready for production deployment?

srallen commented:
WIP because there are a couple of broken tests.

eatyourgreens commented:
To get the tests working with `useContext`, you'll need to wrap your components in a context provider, with a mock context, and use full DOM mounting. Here's an example from the `InteractionLayer` tests:
https://github.com/zooniverse/front-end-monorepo/blob/4a52094b61ba73c81a66d4ba5b8806fb5d00ef7d/packages/lib-classifier/src/components/Classifier/components/SubjectViewer/components/InteractionLayer/InteractionLayer.spec.js#L82-L88

Full mounting might cause problems here, because the Task Nav components are all connected to stores themselves. I remember that being a headache when I first set up the stories and tests for this component.

eatyourgreens commented on packages/lib-classifier/src/components/Classifier/components/TaskArea/components/Tasks/Tasks.js:
```diff
@@ -24,67 +25,74 @@ function storeMapper (stores) {
-    console.error('There was an error loading the workflow steps and tasks.')
-    return (<Paragraph>Something went wrong</Paragraph>)
-  }
+  const stateRenderFunctions = {
```
This object will break if the strings set in `asyncStates` change. I think you can use dynamic object properties to map states to handlers:
```js
function onError() {
…error handling
}
function onSuccess() {
…success stuff.
}
const stateRenderFunctions = {
  [asyncStates.error]: onError,
  [asyncStates.success]: onSuccess
}
```

Alternatively, maybe use dynamic method naming, like the original class did?

```js
const stateRenderFunctions = {
  [asyncStates.error] () {
   …error handling stuff
  },
  [asyncStates.success] () {
   …render the tasks
  }
}
```

srallen replied:
I tried that yesterday but ran into type errors. I can try again.

eatyourgreens commented:
> I was encountering this error when trying to use the knobs:

The knobs are trying to change loading state, which used to be passed in as a prop but now is passed in via the store-mapper. We could split the `Tasks` component into a container and a view, then run the tests and stories against the view, but I think I ran into problems with that because of the Task Nav buttons being wrapped in `inject/observer`.

eatyourgreens commented:
I've opened #1577 to fix those storybook errors, which turned out to be more involved than I thought it would be. The problem was the `store` object being created anew each time a story function runs. I moved the store outside the stories, and set the story functions to update the store via actions.

srallen commented:
I updated enzyme when I updated Grommet and the `wrappingComponent` options should work for when I need context I think: https://enzymejs.github.io/enzyme/docs/api/shallow.html#shallownode-options--shallowwrapper

eatyourgreens commented:
Sure, but I think the tests for things like loading state will fail, because they use props to pass the mock state into the `Tasks` component.

srallen commented:
Closing this as it'll be easier to rework the merge conflict, my storybook refactor, and the hooks refactor separately.
